src/main/core/support/configuration.rs

```diff
@@ -1,9 +1,517 @@
+#[serde(deny_unknown_fields)]
+pub struct ConfigFileOptions {
+    general: GeneralOptions,
+
+    #[serde(default)]
+    host_defaults: HostDefaultOptions,
+
+    #[serde(default)]
+    experimental: ExperimentalOptions,
+
+    topology: Topology,
+
+    // we use a BTreeMap so that the hosts are sorted by their hostname (useful for determinism)
+    hosts: BTreeMap<String, HostOptions>,
+}
+
+/// Shadow configuration options after processing command-line and configuration file options.
+#[derive(Debug, Clone)]
+pub struct ConfigOptions {
+    general: GeneralOptions,
+
+    experimental: ExperimentalOptions,
+
+    topology: Topology,
+
+    // we use a BTreeMap so that the hosts are sorted by their hostname (useful for determinism)
+    hosts: BTreeMap<String, HostOptions>,
+}
+
+impl ConfigOptions {
+    pub fn new(config_file: &ConfigFileOptions, options: &CliOptions) -> Self {
+        let mut config_file = config_file.clone();
+
+        // override config options with command line options
+        config_file.general = GeneralOptions::merge(&options.general, &config_file.general);
+        config_file.host_defaults =
+            HostDefaultOptions::merge(&options.host_defaults, &config_file.host_defaults);
+        config_file.experimental =
+            ExperimentalOptions::merge(&options.experimental, &config_file.experimental);
+
+        // copy the host defaults to all of the hosts
+        for (_, host) in &mut config_file.hosts {
+            host.options = HostDefaultOptions::merge(&host.options, &config_file.host_defaults);
+        }
+
+        Self {
+            general: config_file.general.clone(),
+            experimental: config_file.experimental.clone(),
+            topology: config_file.topology.clone(),
+            hosts: config_file.hosts.clone(),
+        }
+    }
+}
+
+/// Help messages used by Clap for command line arguments, combining the doc string with
+/// the Serde default.
+static GENERAL_HELP: Lazy<std::collections::HashMap<String, String>> =
+    Lazy::new(|| generate_help_strs(schema_for!(GeneralOptions)));
+
+// these must all be Option types since they aren't required by the CLI, even if they're
+// required in the configuration file
+#[derive(Debug, Clone, Clap, Serialize, Deserialize, Merge, JsonSchema)]
+#[clap(help_heading = "GENERAL (Override configuration file options)")]
+#[serde(deny_unknown_fields)]
+pub struct GeneralOptions {
+    /// The simulated time at which simulated processes are sent a SIGKILL signal
+    #[clap(long, value_name = "seconds")]
+    #[clap(about = GENERAL_HELP.get("stop_time").unwrap())]
+    stop_time: Option<units::Time<units::TimePrefixUpper>>,
+
+    /// Environment variables passed to all simulated processes (ex: "ENV_A=1;ENV_B=2")
+    #[clap(long, value_name = "env-string")]
+    #[clap(about = GENERAL_HELP.get("environment").unwrap())]
+    environment: Option<String>,
+
+    /// Initialize randomness for each thread using seed N
```

Comment:
This makes it sound like every thread gets the same seed, which I think is not true. I think it's some kind of `f(seed, threadId)`?

Reply:
I'm not sure what it does, for some of these I just copied the existing help text over. I've changed this now to just say "Initialize randomness using seed N".